### ISSUE TITLE
Receiving error "error for tree [null]

### ISSUE THREAD

**Issue by msmafra** [restic/restic#3151]
Output of `restic version`

❯ restic version
`restic 0.11.0 compiled with go1.15.3 on linux/amd64`

How did you run restic exactly?
From my script, backup works but stats and prune don't.
```
export RESTIC_REPOSITORY="/run/media/marcelo/THEGREY/BACKUPS/tars_linux/" && export RESTIC_PASSWORD_COMMAND="\gpg --quiet --decrypt /home/marcelo/Projects/rest-o/pwd.restic.txt.gpg" 
\restic backup --repo "${RESTIC_REPOSITORY}" --force --verbose --exclude-file="${RESTIC_EXCLUDED:-}" --files-from "${RESTIC_INCLUDED}" "${RESTIC_USER_HOME:-}"
restic prune [breaks]
restic stats [breaks]
```

What backend/server/service did you use to store the repository?
`export RESTIC_REPOSITORY="/run/media/marcelo/THEGREY/BACKUPS/tars_linux/" && export RESTIC_PASSWORD_COMMAND="\gpg --quiet --decrypt /home/marcelo/Projects/rest-o/pwd.restic.txt.gpg" `

Expected behaviour
Removal of unneeded data with the usage of prune command and get statistics from the repository with the stats command.
The check command is returning various blobs errors

Actual behaviour
```
❯ restic prune --verbose 3
repository 6d9f087d opened successfully, password is correct
counting files in repo
building new index for repo
[0:00] 100.00%  11878 / 11878 packs
repository contains 11878 packs (408993 blobs) with 58.120 GiB
processed 408993 blobs: 0 duplicate blobs, 0 B duplicate
load all snapshots
find data that is still in use for 51 snapshots
[0:21] 64.71%  33 / 51 snapshots
id 0000000000000000000000000000000000000000000000000000000000000000 not found in repository
github.com/restic/restic/internal/repository.(*Repository).LoadBlob
	/restic/internal/repository/repository.go:156
github.com/restic/restic/internal/repository.(*Repository).LoadTree
	/restic/internal/repository/repository.go:723
github.com/restic/restic/internal/restic.FindUsedBlobs
	/restic/internal/restic/find.go:19
main.getUsedBlobs
	/restic/cmd/restic/cmd_prune.go:276
main.pruneRepository
	/restic/cmd/restic/cmd_prune.go:158
main.runPrune
	/restic/cmd/restic/cmd_prune.go:62
main.glob..func19
	/restic/cmd/restic/cmd_prune.go:27
github.com/spf13/cobra.(*Command).execute
	/home/build/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826
github.com/spf13/cobra.(*Command).ExecuteC
	/home/build/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914
github.com/spf13/cobra.(*Command).Execute
	/home/build/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main
	/restic/cmd/restic/main.go:98
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374
```

```
❯ restic stats
repository 6d9f087d opened successfully, password is correct
scanning...
error walking snapshot: walking tree 0000000000000000000000000000000000000000000000000000000000000000: id 0000000000000000000000000000000000000000000000000000000000000000 not found in repository

```

```
❯ restic check
using temporary cache in /tmp/restic-check-cache-801153215
repository 6d9f087d opened successfully, password is correct
created new cache in /tmp/restic-check-cache-801153215
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
error for tree [null]:
  id 0000000000000000000000000000000000000000000000000000000000000000 not found in repository
Fatal: repository contains errors
```
_restic check --check-unused_ finds various blobs, more than my terminal can show me without a less/more piped. And issues an error at the end:
...
```
unused blob <data/63a863fd>
unused blob <data/66d14b00>
unused blob <data/dcc78d75>
unused blob <tree/20c594fa>
unused blob <data/158204f3>
unused blob <data/3ff78ad6>
unused blob <data/7e6034c5>
unused blob <data/d3110c55>
unused blob <data/9a1e75c9>
unused blob <tree/49b7fcff>
unused blob <data/3c41fe3e>
unused blob <data/5420f020>
unused blob <data/fab494e9>
unused blob <data/39e39119>
unused blob <data/feb2cec1>
unused blob <data/bcd2e562>
unused blob <data/cdc2d274>
unused blob <data/36e113dd>
unused blob <data/a10905b9>
unused blob <tree/9331a7b5>
unused blob <data/3ecaa4e5>
unused blob <data/56769a41>
unused blob <data/685b08db>
unused blob <data/c34fdde0>
unused blob <data/06dc8a71>
unused blob <tree/ef06d66d>
Fatal: repository contains errors
```

Steps to reproduce the behaviour
Do a backup, run stats, run prune, run check
Running from my script or manually the backup command works. The mount, find, rebuild-index commands also work.
Mounting and copying some random files from different snapshots from different months worked.

Do you have any idea what may have caused this?
Not really. Just noticed it after updating to 0.11.0. I use a script to back up, prune, check and get the stats before and after the backup, but I do not check the report every time. I think I checked once when I updated to 0.10. Running it manually also shows the errors. I checked both my system filesystem and the external USB 3 HDD, both with BTRFS. Check with _btrfs scrub_ and _btrfs filesystem usage_

Do you have an idea how to solve the issue?
No

Did restic help you today? Did it make you happy in any way?
...

**Comment by MichaelEischer** [comment]
Thanks for the detailed description. The error looks like something went slightly wrong during backup and restic ended up creating a snapshot with a null tree instead of failing. This is not a problem of the underlying repository storage as that would cause restic to complain about broken packs etc.

Could you run the following command to try to find out which snapshot is affected?
```
restic list snapshots -q | while read idxid; do \
        restic cat -q snapshot $idxid | jq 'select(.tree | contains ("0000000000000000000000000000000000000000000000000000000000000000"))'; \
done
```

The most interesting data there is probably when the snapshot was created, because based on that information we might find out which restic version was in use at that point in time.

**Comment by msmafra** [comment]
> Could you run the following command to try to find out which snapshot is affected?

Here it is:
```
❯ restic list snapshots -q | while read idxid; do \
        restic cat -q snapshot $idxid | jq 'select(.tree | contains ("0000000000000000000000000000000000000000000000000000000000000000"))'; \
done

{
  "time": "2020-11-15T22:36:24.252957929-03:00",
  "tree": "0000000000000000000000000000000000000000000000000000000000000000",
  "paths": [
    "/home/marcelo",
    "/home/marcelo/Games/steamapps/compatdata/275850/pfx/drive_c/users/steamuser/Application Data/HelloGames/NMS"
  ],
  "hostname": "tars",
  "username": "marcelo",
  "uid": 1000,
  "gid": 1000
}

```

**Comment by msmafra** [comment]
Removing it seems to have solved the problem:

```
❯ restic forget 9f16b858
repository 6d9f087d opened successfully, password is correct
[0:00] 100.00%  1 / 1 files deleted
❯ restic stats
repository 6d9f087d opened successfully, password is correct
scanning...
Stats in restore-size mode:
Snapshots processed:   50
   Total File Count:   15520894
         Total Size:   57.714 GiB
❯ restic check
using temporary cache in /tmp/restic-check-cache-129102296
repository 6d9f087d opened successfully, password is correct
created new cache in /tmp/restic-check-cache-129102296
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
no errors were found

❯ restic prune --verbose 3
repository 6d9f087d opened successfully, password is correct
counting files in repo
building new index for repo
[4:03] 100.00%  11878 / 11878 packs
repository contains 11878 packs (408993 blobs) with 58.120 GiB
processed 408993 blobs: 0 duplicate blobs, 0 B duplicate
load all snapshots
find data that is still in use for 50 snapshots
[0:24] 100.00%  50 / 50 snapshots
found 393540 of 408993 data blobs still in use, removing 15453 blobs
will remove 0 invalid files
will delete 1348 packs and rewrite 686 packs, this frees 8.788 GiB
[2:06] 100.00%  686 / 686 packs rewritten
counting files in repo
[2:16] 100.00%  10219 / 10219 packs
finding old index files
saved new indexes as [8a7b0197 213e5495 ad585f40 5c38162c]
remove 5 old index files
[0:00] 100.00%  5 / 5 files deleted
remove 2034 old packs
[0:00] 100.00%  2034 / 2034 files deleted
done

❯ restic check --check-unused
using temporary cache in /tmp/restic-check-cache-072009413
repository 6d9f087d opened successfully, password is correct
created new cache in /tmp/restic-check-cache-072009413
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
no errors were found
```
Rebuilding index seems to be producing different results every execution
❯ restic rebuild-index --verbose=3
repository 6d9f087d opened successfully, password is correct
counting files in repo
[0:00] 100.00%  10219 / 10219 packs
finding old index files
saved new indexes as [536a817e 75b94a63 bb2189ad 10eb9f6b]
remove 4 old index files
removed <index/5c38162c60>
removed <index/8a7b019749> deleted
removed <index/ad585f402e>
removed <index/213e5495f2>
[0:00] 100.00%  4 / 4 files deleted
❯ restic rebuild-index --verbose=3
repository 6d9f087d opened successfully, password is correct
counting files in repo
[0:00] 100.00%  10219 / 10219 packs
finding old index files
saved new indexes as [200d54bd 1881dc38 c8cacd6c e16effb7]
remove 4 old index files
removed <index/10eb9f6b01>
removed <index/bb2189ad01> deleted
removed <index/75b94a6329>
removed <index/536a817e5b>
[0:00] 100.00%  4 / 4 files deleted
❯ restic rebuild-index --verbose=3
repository 6d9f087d opened successfully, password is correct
counting files in repo
[0:00] 100.00%  10219 / 10219 packs
finding old index files
saved new indexes as [0a1b4cf1 ead54fa1 92dde267 484a299b]
remove 4 old index files
removed <index/1881dc3852>
removed <index/200d54bd3f>
removed <index/e16effb78a> deleted
removed <index/c8cacd6c51>
[0:00] 100.00%  4 / 4 files deleted


```

**Comment by fd0** [comment]
Thanks for the feedback! It looks like restic created an empty snapshot when you ran it on 2020-11-15. Do you maybe still have an idea what you did on that particular date and which version of restic that was? :stuck_out_tongue_winking_eye: Did you cancel restic with CTRL+C?

I remember there was a bug some time ago which made restic create a snapshot although some fatal error occurred, but I can't seem to find it...

**Comment by fd0** [comment]
Someone else also reported the same issue in the forum: https://forum.restic.net/t/invalid-repo-id-00-00-not-found-in-repository/3348/4

Their version was also 0.11.0 (although on Windows). I conclude we have some kind of issue with early cancellation in restic...

**Comment by fd0** [comment]
I'm able to reproduce it, I'll add details later

**Comment by fd0** [comment]
Add the following patch:

```diff
diff --git cmd/restic/cleanup.go cmd/restic/cleanup.go
index 738ec590..8246f81a 100644
--- cmd/restic/cleanup.go
+++ cmd/restic/cleanup.go
@@ -5,6 +5,7 @@ import (
    "os/signal"
    "sync"
    "syscall"
+   "time"

    "github.com/restic/restic/internal/debug"
 )
@@ -76,5 +77,6 @@ func CleanupHandler(c <-chan os.Signal) {
 // given exit code.
 func Exit(code int) {
    RunCleanupHandlers()
+   time.Sleep(10 * time.Second)
    os.Exit(code)
 }
diff --git internal/archiver/archiver.go internal/archiver/archiver.go
index 44081812..86c3996c 100644
--- internal/archiver/archiver.go
+++ internal/archiver/archiver.go
@@ -786,6 +786,9 @@ func (arch *Archiver) Snapshot(ctx context.Context, targets []string, opts Snaps
    start := time.Now()

    debug.Log("starting snapshot")
+
+   time.Sleep(5 * time.Second)
+
    rootTreeID, stats, err := func() (restic.ID, ItemStats, error) {
        tree, err := arch.SaveTree(wctx, "/", atree, arch.loadParentTree(wctx, opts.ParentSnapshot))
        if err != nil {
```

Run a backup, directly cancel it (so the global context is cancelled). This makes `arch.Snapshot()` return `context.Canceled`, which is ignored in the current code. Then, it cancels the derived context, waits for all workers to return, and then ignores the `context.Canceled` error. At the end, a new snapshot is created, with the returned tree with a null ID. That was fun!

I've pushed a fix as #3155, but I'm not sure how complete it is.

**Comment by MichaelEischer** [comment]
There another requirement here: The backend still has to complete `Save()` operations even when the passed context is already canceled. Otherwise it's impossible for the broken snapshot to be saved to the repository.

**Comment by msmafra** [comment]
> Thanks for the feedback! It looks like restic created an empty snapshot when you ran it on 2020-11-15. Do you maybe still have an idea what you did on that particular date and which version of restic that was?  Did you cancel restic with CTRL+C?
> 
> I remember there was a bug some time ago which made restic create a snapshot although some fatal error occurred, but I can't seem to find it...

CTRL + C can definitely have happened. I'm clumsy.

**Comment by msmafra** [comment]
> Add the following patch:
> 
> ```diff
> diff --git cmd/restic/cleanup.go cmd/restic/cleanup.go
> index 738ec590..8246f81a 100644
> --- cmd/restic/cleanup.go
> +++ cmd/restic/cleanup.go
> @@ -5,6 +5,7 @@ import (
>     "os/signal"
>     "sync"
>     "syscall"
> +   "time"
> 
>     "github.com/restic/restic/internal/debug"
>  )
> @@ -76,5 +77,6 @@ func CleanupHandler(c <-chan os.Signal) {
>  // given exit code.
>  func Exit(code int) {
>     RunCleanupHandlers()
> +   time.Sleep(10 * time.Second)
>     os.Exit(code)
>  }
> diff --git internal/archiver/archiver.go internal/archiver/archiver.go
> index 44081812..86c3996c 100644
> --- internal/archiver/archiver.go
> +++ internal/archiver/archiver.go
> @@ -786,6 +786,9 @@ func (arch *Archiver) Snapshot(ctx context.Context, targets []string, opts Snaps
>     start := time.Now()
> 
>     debug.Log("starting snapshot")
> +
> +   time.Sleep(5 * time.Second)
> +
>     rootTreeID, stats, err := func() (restic.ID, ItemStats, error) {
>         tree, err := arch.SaveTree(wctx, "/", atree, arch.loadParentTree(wctx, opts.ParentSnapshot))
>         if err != nil {
> ```
> 
> Run a backup, directly cancel it (so the global context is cancelled). This makes `arch.Snapshot()` return `context.Canceled`, which is ignored in the current code. Then, it cancels the derived context, waits for all workers to return, and then ignores the `context.Canceled` error. At the end, a new snapshot is created, with the returned tree with a null ID. That was fun!
> 
> I've pushed a fix as #3155, but I'm not sure how complete it is.

Sorry, should I try it? Not sure how to do that.

**Comment by fd0** [comment]
> Sorry, should I try it? Not sure how to do that.

Oh, no, sorry! that was just for the other developers reading along, it's my way to locally reproduce the issue, so it's not a solution but quite the opposite ;)
